src/dhtproto/client/legacy/internal/helper/Mirror.d

```diff
@@ -0,0 +1,533 @@
+
+    override protected void assignGetAll ( )
+    {
+        this.allocateGetAlls();
+
+        foreach ( addr, get_all; this.get_alls )
+        {
+            get_all.one_shot = false;
+            get_all.start();
+        }
+    }
+
+    /***************************************************************************
+
+        Schedules a GetAll request that will periodically repeat, once it has
+        finished.
+
+    ***************************************************************************/
+
+    override protected void scheduleGetAll ( )
+    {
+        this.allocateGetAlls();
+
+        foreach ( addr, get_all; this.get_alls )
+        {
+            get_all.one_shot = false;
+            get_all.start(this.update_time_ms);
+        }
+    }
+
+    /***************************************************************************
+
+        Iterates over the set of nodes in the client's registry and creates one
+        Listen object for each. If an instance alreafy exists for a node, it is
+        not recreated.
+
+    ***************************************************************************/
+
+    private void allocateListens ( )
+    {
+        foreach ( node; this.dht.nodes )
+        {
+            auto addr = NodeItem(node.address, node.port);
+            if ( !(addr in this.listens) )
+                this.listens[addr] = new Listen(addr);
+        }
+    }
+
+    /***************************************************************************
+
+        Iterates over the set of nodes in the client's registry and creates one
+        GetAll object for each. If an instance alreafy exists for a node, it is
+        not recreated.
+
+    ***************************************************************************/
+
+    private void allocateGetAlls ( )
+    {
+        foreach ( node; this.dht.nodes )
+        {
+            auto addr = NodeItem(node.address, node.port);
+            if ( !(addr in this.get_alls) )
+                this.get_alls[addr] = new GetAll(addr);
+        }
+    }
+
+    /***************************************************************************
+
+        Called by GetAll.notifier when a GetAll request finishes successfully.
+        If all GetAll instances have finished successfully at least once, calls
+        fillFinished.
+
+    ***************************************************************************/
+
+    private void getAllSucceeded ( )
+    {
+        foreach ( addr, get_all; this.get_alls )
+            if ( !get_all.succeeded() )
+                return;
+
+        this.fillFinished();
+    }
+
+    /***************************************************************************
+
+        GetAll finished notification.
+
+        The base class does nothing, but this method may be overridden by
+        derived classes to implement special fill behaviour.
+
+    ***************************************************************************/
+
+    protected void fillFinished ( )
+    {
+    }
```

Comment:
Because the base class has no way of knowing when this method should be called, and no way of enforcing that derived classes must call it when appropriate.

Reply:
Fair enough.  I guess it's reasonable that anyone using an actual `MirrorBase` implementation knows which particular subclass they are using, and can rely on its extra features (i.e. one would not use `MirrorBase` directly in an app).